core/modal.c

```diff
@@ -0,0 +1,371 @@
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/**
+ * Runtime infrastructure for modes in the C target of Lingua Franca.
+ * This file contains functions that handle the processing of mode transitions.
+ * It works together with other changes across the runtime implementation.
+ *
+ * Any mode related code will only work in the presence of the MODAL_REACTORS
+ * definition.
+ *
+ * This file is intended for direct include in reactor_common.c.
+ *
+ * @author{Alexander Schulz-Rosengarten <als@informatik.uni-kiel.de>}
+ * @author{Soroush Bateni <soroush@utdallas.edu}
+ */
+#ifdef MODAL_REACTORS
+
+#include "reactor.h"
+
+// Forward declaration of functions and variables supplied by reactor_common.c
+void _lf_trigger_reaction(reaction_t* reaction, int worker_number);
+event_t* _lf_create_dummy_events(trigger_t* trigger, instant_t time, event_t* next, microstep_t offset);
+extern reaction_t** _lf_startup_reactions;
+extern int _lf_startup_reactions_size;
+extern pqueue_t* event_q;
+
+/**
+ * Checks whether the given mode is currently considered active.
+ * This includes checking all enclosing modes.
+ * If any of those is inactive, then so is this one.
+ *
+ * @param mode The mode instance to check.
+ */
+bool _lf_mode_is_active(reactor_mode_t* mode) {
+    /*
+     * This code could be optimized by introducing a cached activity indicator
+     * in all mode states. But for now: no premature optimization.
+     */
+    if (mode != NULL) {
+        //DEBUG_PRINT("Checking mode state of %s", mode->name);
+        reactor_mode_state_t* state = mode->state;
+        while (state != NULL) {
+            // If this or any parent mode is inactive, return inactive
+            if (state->active_mode != mode) {
+                //DEBUG_PRINT(" => Mode is inactive");
+                return false;
+            }
+            mode = state->parent_mode;
+            if (mode != NULL) {
+                state = mode->state;
+            } else {
+                state = NULL;
+            }
+        }
+        //DEBUG_PRINT(" => Mode is active");
+    }
+    return true;
+}
+
+/**
+ * Handle startup reactions in modal reactors.
```

Comment:
```suggestion
 * Trigger startup reactions in modal reactors that should be triggered.
 * A startup reaction should be triggered if it is in an active mode and
 * either it has never been triggered before or it has not been triggered
 * since a reset transition was taken into the mode.
```